util/types/src/core/views.rs

```diff
@@ -0,0 +1,662 @@
+            hash: self.hash(),
+        }
+    }
+
+    pub fn fake_hash(mut self, hash: packed::Byte32) -> Self {
+        self.hash = hash;
+        self
+    }
+
+    pub fn calc_proposals_hash(&self) -> H256 {
+        self.data().as_reader().calc_proposals_hash()
+    }
+}
+
+impl UncleBlockVecView {
+    define_simple_getter!(data, UncleBlockVec);
+    define_simple_getter!(hashes, Byte32Vec);
+
+    pub fn get(&self, index: usize) -> Option<UncleBlockView> {
+        if index >= self.data().len() {
+            None
+        } else {
+            Some(self.get_unchecked(index))
+        }
+    }
+
+    pub fn get_unchecked(&self, index: usize) -> UncleBlockView {
+        let data = self.data().get(index).should_be_ok();
+        let hash = self.hashes().get(index).should_be_ok();
+        UncleBlockView { data, hash }
+    }
+}
+
+pub struct UncleBlockVecViewIterator(UncleBlockVecView, usize, usize);
+
+impl ::std::iter::Iterator for UncleBlockVecViewIterator {
+    type Item = UncleBlockView;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.1 >= self.2 {
+            None
+        } else {
+            let index = self.1;
+            self.1 += 1;
+            let data = self.0.data().get(index).should_be_ok();
+            let hash = self.0.hashes().get(index).should_be_ok();
+            Some(UncleBlockView { data, hash })
+        }
+    }
+}
+
+impl ::std::iter::ExactSizeIterator for UncleBlockVecViewIterator {
+    fn len(&self) -> usize {
+        self.2 - self.1
+    }
+}
+
+impl ::std::iter::IntoIterator for UncleBlockVecView {
+    type Item = UncleBlockView;
+    type IntoIter = UncleBlockVecViewIterator;
+    fn into_iter(self) -> Self::IntoIter {
+        let len = self.data().len();
+        UncleBlockVecViewIterator(self, 0, len)
+    }
+}
+
+macro_rules! define_block_unpacked_inner_getter {
+    ($field:ident, $type:ident) => {
+        pub fn $field(&self) -> $type {
+            self.data().as_reader().header().raw().$field().unpack()
+        }
+    }
+}
+
+impl BlockView {
+    define_simple_getter!(data, Block);
+    define_simple_getter!(hash, Byte32);
+    define_simple_getter!(uncle_hashes, Byte32Vec);
+
+    define_vector_getter!(tx_hashes, Byte32);
+    define_vector_getter!(tx_witness_hashes, Byte32);
+
+    define_block_unpacked_inner_getter!(version, Version);
+    define_block_unpacked_inner_getter!(number, BlockNumber);
+    define_block_unpacked_inner_getter!(difficulty, U256);
+    define_block_unpacked_inner_getter!(timestamp, u64);
+    define_block_unpacked_inner_getter!(parent_hash, H256);
+    define_block_unpacked_inner_getter!(transactions_root, H256);
+    define_block_unpacked_inner_getter!(witnesses_root, H256);
+    define_block_unpacked_inner_getter!(proposals_hash, H256);
+    define_block_unpacked_inner_getter!(uncles_hash, H256);
+    define_block_unpacked_inner_getter!(uncles_count, u32);
+    define_block_unpacked_inner_getter!(epoch, EpochNumber);
```

Comment:
@quake

Only `HeaderView`, `TransactionView` and `BlockView` have direct methods to access the inner data.
(And those data was unpacked, since those data were always required `unpack` version when used, if u want a packed version, I think u have to write like `block.data().xx().xx().xx()`).

When we remove `H256`, and only keep `Byte32`, most of these methods will return packed version data, too.

IMHO, it's unnecessarily for other structs. Too many methods is troublesome, too.

For example, a lot of `new` methods, no one can understand them, unless them jump to the definitions.
And we even have a `new` method has 8 inputs, it's not easy to write them in a right order, I think `builder` is better.

Reply:
I use macros because I was really tired to distinguish functions of those methods.

FOr example: a struct has eight fields, there are eight methods named as the field names: seven of them is `getter`, and only one has logical process and that one method return a processed data, not the original data. 
That sucks!